cpp_common/src/lib.rs

```diff
@@ -114,243 +170,115 @@ impl Class {
-                    if let syn::NestedMetaItem::MetaItem(MetaItem::Word(ref d)) = y {
-                        d.as_ref() == i
-                    } else {
-                        false
-                    }
-                })
-            } else {
-                false
-            }
+            use syn::{Meta, NestedMeta};
+            x.interpret_meta().map_or(false, |m| {
+                if let Meta::List(ref list) = m {
+                    list.ident == "derive" && list.nested.iter().any(|y| {
+                        if let NestedMeta::Meta(Meta::Word(ref d)) = y {
+                            d == i
+                        } else {
+                            false
+                        }
+                    })
+                } else {
+                    false
+                }
+            })
         })
     }
 }
 
+impl Parse for Class {
+    fn parse(input: ParseStream) -> Result<Self> {
+        Ok(Class {
+            attrs: input.call(Attribute::parse_outer)?,
+            name: {
+                input.parse::<syn::Visibility>()?;
+                input.parse::<Token![unsafe]>()?;
+                input.parse::<Token![struct]>()?;
+                input.parse()?
+            },
+            cpp: {
+                input.parse::<Token![as]>()?;
+                input.parse::<syn::LitStr>()?.value()
+            },
+            line: String::new(),
+        })
+    }
+}
+
+#[derive(Debug)]
 pub enum Macro {
     Closure(Closure),
-    Lit(Spanned<String>),
+    Lit(TokenStream),
+}
+
+impl Parse for Macro {
+    fn parse(input: ParseStream) -> Result<Self> {
+        if input.peek(syn::token::Brace) {
+            let content;
+            braced!(content in input);
+            return Ok(Macro::Lit(content.parse()?));
+        }
+        Ok(Macro::Closure(input.parse::<Closure>()?))
+    }
 }
 
+#[derive(Debug)]
 pub struct RustInvocation {
-    pub begin: usize,
-    pub end: usize,
+    pub begin: Span,
+    pub end: Span,
     pub id: Ident,
     pub return_type: Option<String>,
-    pub arguments: Vec<(String, String)>, // Vec of name and type
+    pub arguments: Vec<(Ident, String)>, // Vec of name and type
 }
 
-pub mod parsing {
-    use super::{Capture, Class, Closure, ClosureSig, Macro, RustInvocation};
-    use syn::parse::{ident, lit, string, tt, ty};
-    use syn::{Ident, MetaItem, NestedMetaItem, Spanned, Ty, DUMMY_SPAN};
-    use synom::space::{block_comment, whitespace};
-
-    macro_rules! mac_body {
-        ($i: expr, $submac:ident!( $($args:tt)* )) => {
-            delimited!(
-                $i,
-                alt!(punct!("[") | punct!("(") | punct!("{")),
-                $submac!($($args)*),
-                alt!(punct!("]") | punct!(")") | punct!("}"))
-            )
-        };
-        ($i:expr, $e:expr) => {
-            mac_body!($i, call!($e))
+impl Parse for RustInvocation {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let rust_token: Ident = input.parse()?;
```

Comment:
Try using the new `custom_keyword!` here to make this cleaner :-)

Reply:
Right,  that did not exist when i did the port :-).